File: Hash-Table/hash_table.rb
class HashTable
  # Implementation of Hash Table with Linear Probing
  def initialize(size = 10)
    @size = size
    @items = Array.new(size)
  end

  def add(key, value)
    hashed_key = hash(key)
    if exists?(key)
      pointer = hashed_key + 1
      puts "Collision at #{hashed_key}. Pointer set to #{pointer}"
      while hashed_key != pointer
        if @items[pointer].nil?
          @items[pointer] = [key, value]
          break
        elsif pointer >= @size
          puts "Pointer set to zero"
          pointer = 0
        else
          pointer += 1
          puts "New pointer at #{pointer}"
        end
      end
      throw "Hash Table full" if hashed_key.eql?(pointer)
    else
      @items[hashed_key] = [key,value]
    end
  end

  def get(key)
    hashed_key = hash(key)
    if exists?(key)
      value = @items[hashed_key].last
      key_in_hash = @items[hashed_key].first

      return value if key_in_hash == key && value != -1

      pointer = hashed_key + 1
      puts "Pointer set to #{pointer}"
      while hashed_key != pointer
        return nil if @items[pointer].nil?
        
        value = @items[pointer].last
        key_in_hash = @items[pointer].first

        if key_in_hash == key && value != -1
          puts "Keys: #{key_in_hash} - #{key}"
          puts "Value found at #{pointer}"
          return value
        end

        if pointer == @size
          puts "Pointer set to zero"
          pointer = 0
        else
          return nil if value == nil
          pointer += 1
          puts "New pointer at #{pointer}"
        end
      end
    end
    puts "Nothing found"
    nil
  end

  def remove(key)
    hashed_key = hash(key)
    @items[hashed_key] = [key, -1]
  end

  def exists?(key)
    hashed_key = hash(key)
    
    item = @items[hashed_key]

    return false if item == nil
    return item.last != -1
  end

  def list
    @items.each_with_index do |item, index|
      puts "(#{index}, #{item})"
    end
  end

  private

  def hash(key)
    key.hash % @size
  end
end

File: Hash-Table/C#/HashTable.cs
using System.Collections.Generic;

namespace Hash_Tables
{
    /// <summary>
    /// CLass describing a hash table using linear probing.
    /// </summary>
    public class HashTable
    {
        private readonly int _size;
        private readonly Hash[] _hash;

        /// <summary>
        /// Initialises a new instance of the <see cref="HashTable"/> class.
        /// </summary>
        /// <param name="size">The size.</param>
        public HashTable(int size = 10)
        {
            _size = size;
            _hash = new Hash[size];

            for (int i = 0; i < size; i++)
            {
                _hash[i] = new Hash();
            }
        }

        /// <summary>
        /// Adds the specified key and value to the hash table.
        /// </summary>
        /// <param name="key">The key.</param>
        /// <param name="value">The value.</param>
        public void Add(int key, string value)
        {
            var hashedKey = GetHashFromKey(key);

            if (!Exists(key))
            {
                _hash[GetHashFromKey(key)] = new Hash(key, value);
                return;
            }

            var pointer = hashedKey != _size ? hashedKey + 1 : 0;
            
            while (pointer != hashedKey)
            {
                var item = _hash[pointer];

                if (item == null || string.IsNullOrEmpty(item.Value))
                {
                    _hash[pointer] = new Hash(key, value);
                    return;
                }

                if (pointer == hashedKey)
                {
                    pointer = 0;
                }
                else
                {
                    pointer++;
                }
            }
        }

        /// <summary>
        /// Check if an element exists in the underlying array.
        /// </summary>
        /// <param name="key">The key.</param>
        /// <returns></returns>
        public bool Exists(int key)
        {
            var hashedKey = GetHashFromKey(key);
            var item = _hash[GetHashFromKey(key)];

            return !string.IsNullOrEmpty(item.Value);
        }

        /// <summary>
        /// Gets the specified value by its key.
        /// </summary>
        /// <param name="key">The key.</param>
        /// <returns></returns>
        public string Get(int key)
        {
            var hashedKey = GetHashFromKey(key);

            if (!Exists(key))
            {
                return null;
            }

            Hash item = _hash[hashedKey];

            if (item.Key == key && !string.IsNullOrEmpty(item.Value))
            {
                return item.Value;
            }

            var pointer = hashedKey + 1;

            if (item.Key == key && !string.IsNullOrEmpty(item.Value))
            {
                return item.Value;
            }

            if (pointer == _size)
            {
                pointer = 0;
            }

            while (pointer != hashedKey)
            {
                item = _hash[pointer];

                if (item.Key == key && !string.IsNullOrEmpty(item.Value))
                {
                    return item.Value;
                }

                if (pointer == hashedKey)
                {
                    pointer = 0;
                }
                else
                {
                    pointer++;
                }
            }

            return null;
        }

        /// <summary>
        /// Removes the specified item by its key.
        /// </summary>
        /// <param name="key">The key.</param>
        public void Remove(int key)
        {
            if (!Exists(key))
            {
                return;
            }

            _hash[GetHashFromKey(key)] = new Hash();
        }

        /// <summary>
        /// Returns a <see cref="System.String" /> that represents this instance.
        /// </summary>
        /// <returns>
        /// A <see cref="System.String" /> that represents this instance.
        /// </returns>
        public override string ToString()
        {
            List<string> items = new List<string>();

            foreach (var item in _hash)
            {
                items.Add($"({item.Key},{item.Value})");
            }

            return string.Join(", ", items);
        }


        /// <summary>
        /// Returns a hash code for this instance.
        /// </summary>
        /// <returns>
        /// A hash code for this instance, suitable for use in hashing algorithms and data structures like a hash table. 
        /// </returns>
        private int GetHashFromKey(int key)
        {
            return key.GetHashCode() % _size;
        }
    }
}
